### PR TITLE
Update pyproject.toml and remove setup.py

### PR DESCRIPTION
**Pull request by mavaylon1**
- Update version control to hatch backend
- Update pytest, coverage, and ruff via pyproject.toml
- Add ruff and codespell

## Motivation

What was the reasoning behind this change? Please explain the changes briefly.

## How to test the behavior?
```
Show how to reproduce the new behavior (can be a bug fix or a new feature)
```

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `ruff` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf-zarr/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by mavaylon1** [comment]
Fix #110, #91

**Comment by mavaylon1** [comment]
Note: I built locally and the version is correct. Let's aim for a release to test the real thing. (After other PRs are also merged)

**Comment by oruebel** [comment]
Thanks @mavaylon1. @rly  could you please review this one. I think it looks good but you are more familiar with the versioning and deployment workflows.

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,87 @@
...
+    'numcodecs>=0.9.1',
+    'numcodecs==0.11.0',
+    'pynwb>=2.5.0',
+    'threadpoolctl>=3.1.0',
+]
+dynamic = ["version"]
+
+[project.optional-dependencies]
+tqdm = ["tqdm>=4.41.0"]
+fsspec = ["ffspec"]
```
Comment (rly) [inline pyproject.toml]:
```suggestion
fsspec = ["fspec"]
```

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,87 @@
...
+    'numcodecs>=0.9.1',
+    'numcodecs==0.11.0',
+    'pynwb>=2.5.0',
+    'threadpoolctl>=3.1.0',
+]
+dynamic = ["version"]
+
+[project.optional-dependencies]
+tqdm = ["tqdm>=4.41.0"]
+fsspec = ["fspec"]
```
Comment (rly) [inline pyproject.toml]:
```suggestion
fsspec = ["fsspec"]
```

File: pyproject.toml [hunk 3]
```diff
@@ -25,3 +109,27 @@ force-exclude = '''
...
+  "src/hdmf_zarr/_due.py",
+  "docs/source/tutorials/",
+  "docs/_build/",
+]
+line-length = 120
+
+[tool.ruff.per-file-ignores]
+"docs/gallery/*" = ["E402", "T201"]
+"src/*/__init__.py" = ["F401"]
+"setup.py" = ["T201"]
```
Comment (rly) [inline pyproject.toml]:
```suggestion
```

**Comment by rly** [comment]
Please delete `setup.py`

**Comment by rly** [comment]
Please also delete `setup.cfg`

**Comment by rly** [comment]
Please update the required checks so that flake8 is not required but ruff is.

**Review by rly — DISMISSED** [review]

**Comment by mavaylon1** [comment]
@rly I deleted flake8 can you approve again.

**Review by rly — APPROVED** [review]

**Pull request merged**